experimental/packages/api-logs/src/types/AnyValue.ts

```diff
@@ -14,16 +14,29 @@
  * limitations under the License.
  */
 
-import { AttributeValue } from '@opentelemetry/api';
+export type AnyValueScalar = string | number | boolean;
+
+export type AnyValueArray = Array<AnyValue>;
 
 /**
  * AnyValueMap is a map from string to AnyValue (attribute value or a nested map)
  */
 export interface AnyValueMap {
-  [attributeKey: string]: AnyValue | undefined;
+  [attributeKey: string]: AnyValue;
 }
 
 /**
- * AnyValue is a either an attribute value or a map of AnyValue(s)
+ * AnyValue can be one of the following:
+ * - a scalar value
+ * - a byte array
+ * - array of any value
+ * - map from string to any value
+ * - empty value
  */
-export type AnyValue = AttributeValue | AnyValueMap;
+export type AnyValue =
+  | AnyValueScalar
+  | Uint8Array
```

Comment:
There are several typed array types, do we want to have them all listed here? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray  Maybe having a AnyTypedArray type now

Reply:
I think any of the other TypedArrays would get borked on OTLP export. The `toAnyValue()` in otlp-transformer specifically knows about Uint8Array: https://github.com/open-telemetry/opentelemetry-js/blob/01cea7caeb130142cc017f77ea74834a35d0e8d6/experimental/packages/otlp-transformer/src/common/internal.ts#L48

But doesn't know the other TypeArrays.

Reply:
I followed what the spec wording that said `A byte array,` as something that should be supported as any value.

Since byte is 0-255, I think it aligns with Uint8 in contrast to `Int8Array` for example, which semantically is not byte array I think. as @trentm mentioned, they cannot be serialized [into OTLP as `bytesValue` anyhow](https://github.com/open-telemetry/opentelemetry-proto/blob/9833c48ed163156e492882e35aa1a1263a1f6a01/opentelemetry/proto/common/v1/common.proto#L38) as the proto field only accepts `bytes bytes_value = 7;`.

However, it will still be exported (to my understanding) [as `arrayValue`](https://github.com/open-telemetry/opentelemetry-js/blob/01cea7caeb130142cc017f77ea74834a35d0e8d6/experimental/packages/otlp-transformer/src/common/internal.ts#L49C1-L50C62)

```js
  if (Array.isArray(value))
    return { arrayValue: { values: value.map(toAnyValue) } };
```

Reply:
I think it will be exported as an *object* with integer-string keys, because:

```
> a = new Int8Array([1,2,3])
Int8Array(3) [ 1, 2, 3 ]
> Array.isArray(a)
false
> typeof a
'object'
```

From my playing around on https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2352#issuecomment-2276911752 and with a LogRecord with an attribute `aInt8Array: new Int8Array([-7,8,9])` the OTLP server sees something like this:

```
                KeyValue {
                  key: 'aInt8Array',
                  value: AnyValue {
                    kvlistValue: KeyValueList {
                      values: [
                        KeyValue { key: '0', value: AnyValue { intValue: Long { low: -7, high: -1, unsigned: false } } },
                        KeyValue { key: '1', value: AnyValue { intValue: Long { low: 8, high: 0, unsigned: false } } },
                        KeyValue { key: '2', value: AnyValue { intValue: Long { low: 9, high: 0, unsigned: false } } }
                      ]
                    }
                  }
                },
```

FWIW, this is basically how `JSON.stringify` also serializes these TypedArrays:

```
> JSON.stringify(a)
'{"0":1,"1":2,"2":3}'
```